### PR TITLE
feat(training): launch canonical performance recipes

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary

- let the public training launcher run canonical flat performance recipes with the same `--recipe <function_name>` interface as library recipes; no source flag is required
- use exact exported-recipe membership on the lightweight submission path rather than guessing from the name shape
- preserve recipe-owned model, dataset, topology, batch, precision, dispatcher, CUDA graph, and environment settings
- preserve the performance launch runtime: exact total and per-node GPU validation, rank-local NUMA binding, core Pyxis `srun` flags, offline defaults, GB200 NCCL defaults, and four-GPU-node Slurm segment sizing
- apply recipe environment defaults and re-exec before importing the training loop
- reject unsupported performance workflows before executor construction, while keeping `scripts/performance` as their compatibility launcher

This PR is stacked on #4682.

## Lookup and initial scope

A complete `--recipe` name is discovered automatically. Exact performance exports select the performance path; all other names fall back to the functional recipe library. Three legacy duplicate text-pretrain names are performance-first. Two duplicate SFT/PEFT names remain library-first until those workflows migrate, and regression tests guard the complete collision set.

The unified performance path intentionally supports canonical text pretraining only and uses the canonical `gpt_step` and recipe-owned dataset. Only `train.train_iters`, `logger.*`, `profiling.*`, and `env_vars.*` overrides are accepted so the measured recipe remains canonical.

SFT/PEFT, VLM, diffusion, dataset replacement, topology resizing, and specialized benchmark controls continue to use `scripts/performance/setup_experiment.py` during migration. Unsupported SFT/PEFT, Qwen-VL, and Wan performance recipes are rejected before Slurm submission.

## Validation

- 181 focused training-launcher unit tests passed
- `uv run --no-project --with pre-commit pre-commit run --all-files` passed
- a no-source H100 NeMo-Run 0.10 submission dry run rendered the expected 2x8 allocation, container flags, environment, quoting, and performance runtime
- an unsupported performance SFT command was rejected before executor construction
- canonical Qwen3 30B-A3B configuration and environment were previously validated with a rank-local Linux training-container dry run

A live multi-node performance training run was not submitted as part of this PR validation.

### CODE REVIEW

File: scripts/training/launch_with_nemo_run.py [hunk 1]
```diff
@@ -24,23 +24,24 @@
     # Test locally (single node)
     python launch_with_nemo_run.py \
         --local \
-        --script run_recipe.py \
+        --devices 2 \
         --recipe llama32_1b_pretrain_config \
```
Comment (yaoyu-33) [inline scripts/training/launch_with_nemo_run.py:28]: update to use new names, like h100 how many gpus etc.

File: scripts/training/run_recipe.sh [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#!/usr/bin/env bash
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.sh]: no need this shell.

File: scripts/training/setup_experiment.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+import nemo_run as run
+from nemo_run.config import get_nemorun_home
+
+
+logger = logging.getLogger(__name__)
+
+CONTAINER_REPO_ROOT = Path("/opt/Megatron-Bridge")
+INHERITED_ENV_VARS = (
+    "DCLM_CACHE",
```
Comment (yaoyu-33) [inline scripts/training/setup_experiment.py]: no mention dclm here, also dont see a reason of NEMO_DATASETS_CACHE here. bad design, cannot extend.

File: scripts/training/setup_experiment.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+PATH_ENV_VARS = (
+    "DCLM_CACHE",
+    "DCLM_DATA_DIR",
+    "DCLM_DATA_PREFIX",
+    "HF_HOME",
+    "NEMO_DATASETS_CACHE",
+    "NEMO_HOME",
+    "UV_CACHE_DIR",
+)
+PATH_OPTIONS = (
```
Comment (yaoyu-33) [inline scripts/training/setup_experiment.py]: why these stand alone?

File: scripts/training/setup_experiment.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        help="GPUs per node.",
+    )
+    execution.add_argument("--gpu-type", help="Recipe hardware selector, e.g. h100 or gb200.")
+    execution.add_argument("--account", default=os.environ.get("SLURM_ACCOUNT"), help="Slurm account.")
+    execution.add_argument("--partition", default=os.environ.get("SLURM_PARTITION"), help="Slurm partition.")
+    execution.add_argument("--time", default="04:00:00", help="Slurm time limit.")
+    execution.add_argument("--gres", help="Optional Slurm GRES value.")
+    execution.add_argument(
+        "--container-image",
+        default=os.environ.get("MB_CONTAINER_IMAGE") or os.environ.get("CONTAINER_IMAGE"),
```
Comment (yaoyu-33) [inline scripts/training/setup_experiment.py]: "MB_CONTAINER_IMAGE" no, this is my local use

File: scripts/training/run_recipe.py [hunk 6]
```diff
@@ -12,322 +12,650 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: needs new docstrings

File: scripts/training/run_recipe.py [hunk 7]
```diff
@@ -12,322 +12,650 @@
...
+    """Resolved user-facing dataset selection."""
+
+    dataset_type: str
+    preset: str | None = None
+    packed_sequence: bool = False
+    thinking_format: bool = False
+
+
+DATASET_PRESETS = {
+    "mock": _DatasetSelection("llm-pretrain-mock"),
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: stop use old names like llm-pretrain, llm-finetune etc. if we are switch to dataset name directly just use those names

File: scripts/training/run_recipe.py [hunk 8]
```diff
@@ -12,322 +12,650 @@
...
+        packed_sequence=True,
+        thinking_format=True,
+    ),
+    "gsm8k": _DatasetSelection("llm-finetune", preset="gsm8k"),
 }
 
-TRAIN_FUNCTIONS = {
-    "pretrain": pretrain,
-    "finetune": finetune,
+DATASET_ALIASES = {
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: similar no need.

File: scripts/training/run_recipe.py [hunk 9]
```diff
@@ -12,322 +12,650 @@
...
+        "--etp",
+        "--expert-tensor-parallel-size",
+        "--expert_tensor_parallel_size",
+        type=int,
+        dest="expert_tensor_parallel_size",
+    )
+
+    checkpoint = parser.add_argument_group("Checkpointing")
+    checkpoint.add_argument(
+        "--from",
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: you are doing too many names in options. "-" and "_" 2 cases are fine, but there are more than that. confusing.

File: scripts/training/run_recipe.py [hunk 10]
```diff
@@ -12,322 +12,650 @@
...
+        help="Preset for --dataset llm-finetune.",
     )
-    parser.add_argument(
-        "--seq_length",
+    data.add_argument(
+        "--dataset-path",
+        action="append",
+        default=[],
+        dest="dataset_paths",
+        help="Preprocessed dataset prefix or directory. Repeat for multiple DCLM prefixes.",
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: dont say DCLM.

File: scripts/training/run_recipe.py [hunk 11]
```diff
@@ -12,322 +12,650 @@
...
+        "--config_variant",
+        dest="config_variant",
+        help="Optional non-canonical flat perf recipe variant suffix.",
+    )
+    selection.add_argument(
+        "--mode",
+        choices=["pretrain", "sft", "finetune", "peft", "lora", "dora"],
+        help="User-facing training mode. lora/dora select a PEFT recipe and scheme.",
+    )
+    selection.add_argument(
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py:163]: dont use domain, step func is enough for now. default should be llm_step, most cases no need to pass in. it's okay to break some backward compatibility.

File: scripts/training/run_recipe.py [hunk 12]
```diff
@@ -12,322 +12,650 @@
...
+        "--variant",
+        "--config-variant",
+        "--config_variant",
+        dest="config_variant",
+        help="Optional non-canonical flat perf recipe variant suffix.",
+    )
+    selection.add_argument(
+        "--mode",
+        choices=["pretrain", "sft", "finetune", "peft", "lora", "dora"],
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: "pretrain", "sft", "lora", "dora"; finetune == sft? peft is vague. please fix.

File: scripts/training/run_recipe.py [hunk 13]
```diff
@@ -12,322 +12,650 @@
...
+        "--model",
+        "--model-recipe-name",
+        "--model_recipe_name",
+        "-mr",
+        dest="model_recipe_name",
+        help="Model recipe stem, e.g. llama32_1b, qwen3_8b, deepseek_v3.",
+    )
+    selection.add_argument(
+        "--task",
+        choices=["pretrain", "sft", "finetune", "peft", "lora", "dora"],
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: mode and task? sounds same to me.

File: scripts/training/run_recipe.py [hunk 14]
```diff
@@ -12,322 +12,650 @@
...
+    selection.add_argument(
+        "--source",
+        choices=["auto", "recipes", "perf_recipes"],
+        default="auto",
+        help="Recipe namespace for --recipe or selector mode.",
     )
-    parser.add_argument(
-        "--dataset",
-        type=str,
+    selection.add_argument(
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py]: either --model to pass in model name or --recipe to pass in full recipe name.

update --use-recipes and --model. outdated.

File: scripts/training/run_recipe.py [hunk 15]
```diff
@@ -12,322 +12,650 @@
...
+
+
+def _parse_key_value(value: str) -> str:
+    """Validate KEY=VALUE override syntax for --set."""
+    if "=" not in value:
+        raise argparse.ArgumentTypeError(f"Expected KEY=VALUE, got {value!r}")
+    return value
+
+
+def _build_parser() -> argparse.ArgumentParser:
```
Comment (yaoyu-33) [inline scripts/training/run_recipe.py:146]: just primarily consider one path, do not consider so many corner cases to make things harder in general.

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4620/changes/80054d25c4c89dde8a63af9f0fcaedc0e5b4c0d0
